components/ai_chat/core/browser/engine/engine_consumer_oai.cc

```diff
@@ -389,10 +248,179 @@ base::Value::List EngineConsumerOAIRemote::BuildPageContentMessages(
     messages.Append(std::move(message));
 
     if (truncated_page_content.size() > max_associated_content_length) {
-      break;
+      max_associated_content_length = 0;
     }
     max_associated_content_length -= truncated_page_content.size();
   }
+  return messages;
+}
+
+base::Value::List EngineConsumerOAIRemote::BuildMessages(
+    const mojom::CustomModelOptions& model_options,
+    PageContentsMap& page_contents,
+    std::optional<base::Value::Dict> user_memory_message,
+    const std::optional<std::string>& selected_text,
+    const EngineConsumer::ConversationHistory& conversation_history) {
+  uint32_t remaining_content_length = max_associated_content_length_;
+  uint32_t selected_text_length = selected_text.value_or("").size();
+  if (selected_text_length > max_associated_content_length_) {
+    remaining_content_length = 0;
+  } else {
+    remaining_content_length -= selected_text_length;
+  }
+
+  base::flat_map<std::string, base::Value::List> page_contents_messages;
+
+  // We iterate over the page contents in reverse order so that the most recent
+  // content is preferred.
+  for (size_t i = conversation_history.size(); i > 0; i--) {
+    auto& turn = conversation_history[i - 1];
```

Comment:
Is there a reason that we can't use  `for (const auto& turn : base::Reversed(conversation_history)) {`?

Also I don't see why reversing history will affect page content order when we build that flat_map, it only matters when putting page content into a list in `BuildPageContentMessages` so we don't need an extra reverse iteration through history to build the map first.

Reply:
We need to iterate over the messages in reverse order because we need to prefer content associated with newer messages over content associated with older messages so when we run out of `remaining_content_length` we truncate old content, not new content.

Changed to use `base::Reversed`